### PR TITLE
web3torrent: Add Budget UI

### LINKED ISSUE #1171: [web3t] Add UI for ledger connection to hub
This should probably look a little like the `ChannelsList` component, but 

- pulls the requisite data from a new channel client api call `getBudget` or similar
- has "Withdraw" instead of "Close" button
- shows money not locked in payment channels

reference: 
https://www.notion.so/Web3Torrent-Requirements-ce435f483e044d8291a8a01f570391d4

### PR DESCRIPTION
**Pull request by junjizhi**
Fixes #1171 

Reopens https://github.com/statechannels/monorepo/pull/1298

The budget UI looks like this:

image

### CODE REVIEW

**Comment by junjizhi** [comment]
@geoknee I rebased on master and also wired up the fake channel provider with budget.

File: packages/web3torrent/src/pages/file/File.tsx [hunk 1]
```diff
@@ -53,10 +55,28 @@ const File: React.FC<RouteComponentProps & Props> = props => {
...
+          const showBudget =
+            Object.keys(budgetCache).length > 0 && Object.keys(channelCache).length > 0;
           return (
-            <TorrentInfo torrent={torrent} channelCache={channelCache} mySigningAddress={me} />
+            <>
+              <TorrentInfo torrent={torrent} channelCache={channelCache} mySigningAddress={me} />
+              <br />
+              {showBudget ? (
+                <SiteBudgetTable
+                  budgetCache={budgetCache as SiteBudget}
```
Comment (geoknee) [inline packages/web3torrent/src/pages/file/File.tsx]: Is this type assertion necessary?

File: packages/web3torrent/src/pages/file/File.tsx [hunk 2]
```diff
@@ -53,10 +55,28 @@ const File: React.FC<RouteComponentProps & Props> = props => {
       </div>
       <WebTorrentContext.Consumer>
         {web3Torrent => {
-          const me = web3Torrent.paymentChannelClient.mySigningAddress;
-          const channelCache = web3Torrent.paymentChannelClient.channelCache;
+          const paymentChannelClient = web3Torrent.paymentChannelClient;
+          const me = paymentChannelClient.mySigningAddress;
```
Comment (spersico) [inline packages/web3torrent/src/pages/file/File.tsx]:
```suggestion
          const { paymentChannelClient, mySigningAddress:me, channelCache,budgetCache } = paymentChannelClient;
```

File: packages/web3torrent/src/pages/file/File.tsx (hunk 1 above)
Comment (junjizhi) [reply packages/web3torrent/src/pages/file/File.tsx]: Yeah. We have in the most places defined `SiteBudget | {}` or `SiteBudget | undefined` as the return type.

File: packages/web3torrent/src/components/site-budget-table/SiteBudgetTable.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      key => channelCache[key].beneficiary === mySigningAddress
+    );
+
+    const spent = myPayingChannelIds
+      .map(id => channelCache[id].beneficiaryBalance)
+      .reduce((a, b) => bigNumberify(a).add(bigNumberify(b)), bigNumberify(0));
+
+    const received = myReceivingChannelIds
+      .map(id => channelCache[id].beneficiaryBalance)
+      .reduce((a, b) => bigNumberify(a).add(bigNumberify(b)), bigNumberify(0));
```
[the remark this answers is not in the document]
Comment (junjizhi) [reply packages/web3torrent/src/components/site-budget-table/SiteBudgetTable.tsx]: Good suggestion. I actually tried before but messed up hook dependencies, so the component wasn't updated properly. I think here the calculation is simple enough. I will leave as is now.

**Comment by junjizhi** [comment]
@geoknee 
> One remaining thing is that I think we want this UI to be driven by onBudgetUpdated events. 

Done.

**Pull request merged**